feat: support `/api/v1/series` for Prometheus

NiwakaDev opened the pull request:
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR adds support `/api/v1/series` for Prometheus.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

close #1539

killme2008 commented on src/servers/src/prom.rs:
```diff
@@ -699,3 +741,57 @@ fn promql_expr_to_metric_name(expr: &PromqlExpr) -> Option<String> {
+        let response = retrieve_series_from_query_result(result, &mut series).await;
+
+        if let Err(err) = response {
+            // Prometheus won't report error if querying nonexist label and metric
```
Let's return the errors right now, we will support a compatible mode in future.

NiwakaDev replied:
Do you mean:

```rust
retrieve_series_from_query_result(result, &mut series).await?;
```

killme2008 replied:
Yes, always return the error even when table is not found etc.

killme2008 commented on src/servers/src/prom.rs:
```diff
@@ -607,6 +632,23 @@ async fn retrieve_labels_name_from_query_result(
+) -> Result<()> {
+    for batch in batches.iter() {
+        for row in batch.rows() {
+            let mut element = HashMap::new();
+            for (idx, column) in row.iter().enumerate() {
+                let column_name = batch.schema.column_name_by_index(idx);
+                element.insert(column_name.to_string(), column.to_string());
+            }
```
```suggestion
            let element = row.iter().enumerate()
                    .map(|idx, column| {
                        let column_name = batch.schema.column_name_by_index(idx);
                        (column_name.to_string(), column.to_string())               
                    }).
                    .collect();
```

killme2008 commented on tests-integration/tests/http.rs:
```diff
@@ -336,6 +336,20 @@ pub async fn test_prom_http_api(store_type: StorageType) {
+        .header("Content-Type", "application/x-www-form-urlencoded")
+        .send()
+        .await;
+    assert_eq!(res.status(), StatusCode::OK);
```
Can we assert if the response body is expected?

NiwakaDev replied:
By the way, can I also add tests for `/api/v1/labels` and `/api/v1/label/:label_name/values`?

NiwakaDev replied:
It seems like there aren't the response body tests for `/api/v1/labels` and `/api/v1/label/:label_name/values`

https://github.com/GreptimeTeam/greptimedb/blob/3dc45f1c138a926c70649021bc223b57f1766fba/tests-integration/tests/http.rs#L289-L371

killme2008 replied:
Good catch! Appreciate it if you can add them too. Thanks a lot.

waynexia reviewed (approved):
Other parts look good to me, thanks @NiwakaDev

killme2008 commented:
@NiwakaDev Hello, are u still working on this PR? I think we can merge this PR at first, you can create another PR for the tests.

NiwakaDev commented:
@killme2008 

>are u still working on this PR? 

Yes.

I have encountered a problem where the test code for the `/api/v1/series` is failing to retrieve the expected data, while we can get the correct results using `cargo run -- standalone start`.

## Detail

To test the request body, I modified `setup_test_prom_app_with_frontend`, like so:

```rust
fn mock_insert_request() -> InsertRequest {
    let mut columns_values = HashMap::with_capacity(4);
    let mut builder = StringVectorBuilder::with_capacity(1);
    builder.push(Some("host1"));
    columns_values.insert("host".to_string(), builder.to_vector());

    let mut builder = Float64VectorBuilder::with_capacity(1);
    builder.push(Some(1.1));
    columns_values.insert("cpu".to_string(), builder.to_vector());

    let mut builder = Float64VectorBuilder::with_capacity(1);
    builder.push(Some(2.0));
    columns_values.insert("memory".to_string(), builder.to_vector());

    let mut builder = TimestampSecondVectorBuilder::with_capacity(1);
    builder.push(Some(0.into()));
    columns_values.insert("ts".to_string(), builder.to_vector());

    InsertRequest {
        catalog_name: common_catalog::consts::DEFAULT_CATALOG_NAME.to_string(),
        schema_name: common_catalog::consts::DEFAULT_SCHEMA_NAME.to_string(),
        table_name: "demo".to_string(),
        columns_values,
        region_number: 0,
    }
}

pub async fn setup_test_prom_app_with_frontend(
    store_type: StorageType,
    name: &str,
) -> (Router, TestGuard) {
    let (opts, guard) = create_tmp_dir_and_datanode_opts(store_type, name);
    let instance = Arc::new(Instance::with_mock_meta_client(&opts).await.unwrap());
    let frontend = FeInstance::try_new_standalone(instance.clone())
        .await
        .unwrap();
    instance.start().await.unwrap();
    create_test_table(
        frontend.catalog_manager(),
        instance.sql_handler(),
        ConcreteDataType::timestamp_second_datatype(),
    )
    .await
    .unwrap();

    let table = frontend
        .catalog_manager()
        .table(DEFAULT_CATALOG_NAME, DEFAULT_SCHEMA_NAME, "demo")
        .await
        .unwrap()
        .unwrap();
    let size = table.insert(mock_insert_request()).await.unwrap();
    assert_eq!(size, 1);
    let http_opts = HttpOptions {
        addr: format!("127.0.0.1:{}", ports::get_port()),
        ..Default::default()
    };
    let frontend_ref = Arc::new(frontend);
    let http_server = HttpServerBuilder::new(http_opts)
        .with_sql_handler(ServerSqlQueryHandlerAdaptor::arc(frontend_ref.clone()))
        .with_grpc_handler(ServerGrpcQueryHandlerAdaptor::arc(frontend_ref.clone()))
        .with_script_handler(frontend_ref.clone())
        .with_prom_handler(frontend_ref.clone())
        .build();
    let prom_server = PromServer::create_server(frontend_ref);
    let app = http_server.build(http_server.make_app());
    let app = app.merge(prom_server.make_app());
    (app, guard)
}

```

And then I modified the `test_prom_http_api` method, like so:

```rust
pub async fn test_prom_http_api(store_type: StorageType) {
    common_telemetry::init_default_ut_logging();
    let (app, mut guard) = setup_test_prom_app_with_frontend(store_type, "promql_api").await;
    let client = TestClient::new(app);

    let res = client
        .get("/v1/sql?db=greptime-public&sql=select * from demo")
        .send()
        .await;
    assert_eq!(res.text().await, "{\"code\":0,\"output\":[{\"records\":{\"schema\":{\"column_schemas\":[{\"name\":\"host\",\"data_type\":\"String\"},{\"name\":\"cpu\",\"data_type\":\"Float64\"},{\"name\":\"memory\",\"data_type\":\"Float64\"},{\"name\":\"ts\",\"data_type\":\"TimestampSecond\"}]},\"rows\":[[\"host1\",1.1,2.0,0]]}}],\"execution_time_ms\":1}");

    //...

    // series
    let res = client
        .get("/api/v1/series?match[]=demo&start=0&end=0")
        .header("Content-Type", "application/x-www-form-urlencoded")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);

    assert_eq!(res.text().await, "{\"status\":\"success\",\"data\":[{\"ts\":\"1970-01-01 09:00:00+0000\",\"cpu\":\"1.1\",\"host\":\"host1\",\"memory\":\"2.0\"}]}");
    let res = client
        .post("/api/v1/series?match[]=up&match[]=down")
        .header("Content-Type", "application/x-www-form-urlencoded")
        .send()
        .await;
	
    //...

    guard.remove_all().await;
}
```

The immediately above result is:
```
thread 'integration_http_file_test::test_prom_http_api' panicked at 'assertion failed: `(left == right)`
  left: `"{\"status\":\"success\",\"data\":[]}"`,
 right: `"{\"status\":\"success\",\"data\":[{\"ts\":\"1970-01-01 09:00:00+0000\",\"cpu\":\"1.1\",\"host\":\"host1\",\"memory\":\"2.0\"}]}"`', tests-integration/tests/http.rs:355:5
```

As you can see, we could get the correct result via `/v1/sql?db=greptime-public&sql=select * from demo`, but we couldn't get the correct one via `/api/v1/series?match[]=demo&start=0&end=0`.

Note that we could get the correct result via `/api/v1/series?match[]=demo&start=0&end=0` using `cargo run -- standalone start`.

Am I missing something?

killme2008 commented:
@NiwakaDev Can you push the code? I can help to analyze the issue.

NiwakaDev commented:
@killme2008 

>Can you push the code? I can help to analyze the issue.

I pushed it.

killme2008 commented on tests-integration/src/test_util.rs:
```diff
@@ -338,6 +343,42 @@ pub async fn setup_test_http_app_with_frontend(
+    builder.push(Some(2.0));
+    columns_values.insert("memory".to_string(), builder.to_vector());
+
+    let mut builder = TimestampSecondVectorBuilder::with_capacity(1);
```
I find the cause. It's panic at https://github.com/GreptimeTeam/greptimedb/blob/c5f6d7c99ae25832603be90f0d09bfd20e3337f8/src/promql/src/extension_plan/normalize.rs#L225

Because we suppose the timestamp column type is always `TimestampMillisecond` in Prometheus tables. So the `downcase_ref` will return None and panic when `unwrap`.

NiwakaDev replied:
Thank you for helping me!! I'll write tests today or tomorrow.

killme2008 commented on tests-integration/src/test_util.rs:
```diff
@@ -351,12 +392,33 @@ pub async fn setup_test_prom_app_with_frontend(
         frontend.catalog_manager(),
         instance.sql_handler(),
-        ConcreteDataType::timestamp_millisecond_datatype(),
+        ConcreteDataType::timestamp_second_datatype(),
```
The root cause is here. You can change it back to `timestamp_millisecond_datatype` and fix the test.

killme2008 commented on tests-integration/tests/http.rs:
```diff
@@ -336,6 +354,21 @@ pub async fn test_prom_http_api(store_type: StorageType) {
+
+    let res = client
+        .post("/api/v1/series?match[]=up&match[]=down")
+        .header("Content-Type", "application/x-www-form-urlencoded")
```
The header looks unnecessary.

killme2008 commented on tests-integration/tests/http.rs:
```diff
@@ -336,6 +354,21 @@ pub async fn test_prom_http_api(store_type: StorageType) {
+        .send()
+        .await;
+    assert_eq!(res.status(), StatusCode::OK);
+    assert_eq!(res.text().await, "{\"status\":\"success\",\"data\":[{\"ts\":\"1970-01-01 09:00:00+0000\",\"cpu\":\"1.1\",\"host\":\"host1\",\"memory\":\"2.0\"}]}");
```
I think we can just confirm the response text contains `memory` etc. The column's order in text may not be consistent between tests.

NiwakaDev commented:
@killme2008 , @waynexia 

I apologize for this late reply. 

I found a bug for prom query.
Let's say that we have a `demo1` table, like so:

```rust
    let demo1 = frontend
        .catalog_manager()
        .table(DEFAULT_CATALOG_NAME, DEFAULT_SCHEMA_NAME, "demo1")
        .await
        .unwrap()
        .unwrap();

    let _ = demo1
        .insert(mock_insert_request("host1", 1.1, 2.2, 0))
        .await
        .unwrap();
```

And then I executed some tests:

```rust
    // series
    let res = client
        .get("/api/v1/series?match[]=demo1&start=0&end=300")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    let body = serde_json::from_str::<PromJsonResponse>(&res.text().await).unwrap();
    assert_eq!(body.status, "success");
    assert_eq!(
        body.data,
        serde_json::from_value::<PromResponse>(json!(
            [{"__name__" : "demo1","ts":"1970-01-01 00:00:00+0000","cpu":"1.1","host":"host1","memory":"2.2"}]
        ))
        .unwrap()
    );

```

The above test result is:

```
thread 'integration_http_file_test::test_prom_http_api' panicked at 'assertion failed: `(left == right)`
  left: `Series([{"memory": "2.2", "cpu": "1.1", "host": "host1", "ts": "1970-01-01 00:00:00+0000", "__name__": "demo1"}, {"cpu": "1.1", "memory": "2.2", "__name__": "demo1", "ts": "1970-01-01 00:05:00+0000", "host": "host1"}])`,
 right: `Series([{"host": "host1", "memory": "2.2", "ts": "1970-01-01 00:00:00+0000", "__name__": "demo1", "cpu": "1.1"}])`', tests-integration/tests/http.rs:364:5
```

It seems like the `demo1` table has a row, which has `1970-01-01 00:05:00+0000`. I guess that this is a bug.

In addition, 

```rust
    let res = client
        .get("/api/v1/series?match[]=demo1&start=1&end=1")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    let body = serde_json::from_str::<PromJsonResponse>(&res.text().await).unwrap();
    assert_eq!(body.status, "success");
    assert_eq!(
        body.data,
        serde_json::from_value::<PromResponse>(json!(
            []
        ))
        .unwrap()
    );
```

The immediately above test result is:
```
thread 'integration_http_file_test::test_prom_http_api' panicked at 'assertion failed: `(left == right)`
  left: `Series([{"cpu": "1.1", "ts": "1970-01-01 00:00:01+0000", "__name__": "demo1", "host": "host1", "memory": "2.2"}])`,
 right: `Labels([])`', tests-integration/tests/http.rs:381:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

It seems like the `demo1` table has a row, which has `1970-01-01 00:00:01+0000`. I guess that this is also a bug. 


By the way, we could also get similar response results via `/api/v1/labels`, like so:

```rust
    let res = client
        .get("/api/v1/labels?match[]=demo1&start=1&end=1")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);
    let body = serde_json::from_str::<PromJsonResponse>(&res.text().await).unwrap();
    assert_eq!(body.status, "success");
    assert_eq!(
        body.data,
        serde_json::from_value::<PromResponse>(json!([]))
            .unwrap()
    );
```

result:
```
thread 'integration_http_file_test::test_prom_http_api' panicked at 'assertion failed: `(left == right)`
  left: `Labels(["__name__", "cpu", "host", "memory", "ts"])`,
 right: `Labels([])`', tests-integration/tests/http.rs:349:5
```

## Additional information.

We could get the correct result via `/v1/sql?sql=select * from demo1`: 

```rust
    let res = client
        .get("/v1/sql?sql=select * from demo1")
        .send()
        .await;
    assert_eq!(res.status(), StatusCode::OK);

    let body = serde_json::from_str::<JsonResponse>(&res.text().await).unwrap();
    assert!(body.success());
    assert!(body.execution_time_ms().is_some());
    let output = body.output().unwrap();
    assert_eq!(output.len(), 1);

    assert_eq!(
        output[0],
        serde_json::from_value::<JsonOutput>(json!({
            "records":{"schema":{"column_schemas":[{"name":"host","data_type":"String"},{"name":"cpu","data_type":"Float64"},{"name":"memory","data_type":"Float64"},{"name":"ts","data_type":"TimestampMillisecond"}]},"rows":[["host1",1.1,2.2,0]]}
        })).unwrap()
    );
```

NiwakaDev commented:
@killme2008 , @waynexia 
I pushed [549d315](https://github.com/GreptimeTeam/greptimedb/pull/1620/commits/549d31520b9fab790248057b3bc992eb5fdf7209).

haohuaijin commented:
Hi, @NiwakaDev,
> It seems like the demo1 table has a row, which has 1970-01-01 00:05:00+0000. I guess that this is a bug.

this is not a bug, we default lookback is 5m, and you set `start=0`, `end=300`, step=`300(5m)`, so the result is two row, one row timestamp is in `1970-01-01 00:00:00+0000`, and the other row is in `1970-01-01 00:05:00+0000`, because in timestamp 1970-01-01 00:05:00+0000 don't have value, so prom look back up to 5m to find a value before 1970-01-01 00:05:00+0000. 
> It seems like the demo1 table has a row, which has 1970-01-01 00:00:01+0000. I guess that this is also a bug. 

the timestamp is based on you input `start`,  becuase you input `start=1`, so the first timestamp is `1970-01-01 00:00:01+0000`.

NiwakaDev commented:
@haohuaijin 
Thank you for your quick reply.

>this is not a bug, we default lookback is 5m, and you set start=0, end=300, step=300(5m), so the result is two row, one row's timestamp is in 1970-01-01 00:00:00+0000, and the other row's timestamp is in 1970-01-01 00:05:00+0000, because in timestamp 1970-01-01 00:05:00+0000 don't have value, so prom look back up to 5m to find a value before 1970-01-01 00:05:00+0000.

I see. Thanks!!

killme2008 commented:
@NiwakaDev There is a file conflict have to be resolved.

killme2008 reviewed (approved):
LGTM
